### PR TITLE
ton-o-ui-fixes

### PR DESCRIPTION
**Pull request by tn3rb**
- simplifies and standardizes borders, colours, classes, plus hover and focus states for inputs and buttons

- improves tooltip readability and aligns styles with rest of EDTR

- fixes that damn fugly uber datepicker

- tweaks sidebar calendar elements

### CODE REVIEW

File: assets/src/application/ui/forms/espressoForm/renderers/FieldRenderer.tsx [hunk 1]
```diff
@@ -27,23 +27,27 @@ const FieldRenderer: React.FC<FieldRendererProps> = (props) => {
...
-			<FormLabel htmlFor={ props.input.name }>
-				{ label }
-				{ info && (
-					<Tooltip placement='right' tooltip={ info }>
-						<InfoCircleOutlined />
+			<FormLabel htmlFor={props.input.name}>
+				{label}
+				{info && (
+					<Tooltip placement='right' tooltip={info}>
+						<InfoCircleOutlined className='ee-form-item__help-info' />
```
Comment (manzoorwanijk) [inline assets/src/application/ui/forms/espressoForm/renderers/FieldRenderer.tsx:42]: 👍

File: assets/src/infrastructure/ui/layout/menu/Menu.tsx [hunk 2]
```diff
@@ -4,8 +4,12 @@ import { Menu as ChakraMenu } from '@chakra-ui/core';
 
 import { MenuProps } from './types';
 
-const Menu: React.FC<MenuProps> = ({ children, ...props }) => {
-	return <ChakraMenu {...props}>{children}</ChakraMenu>;
+const Menu: React.FC<MenuProps> = ({ closeOnBlur = true, children, ...props }) => {
```
Comment (nicolad) [inline assets/src/infrastructure/ui/layout/menu/Menu.tsx:7]: set `closeOnBlur` explicitly to `true`, but it's still not closing when it's clicked outside of the area

File: assets/src/application/ui/display/TimezoneTimeInfo/style.scss [hunk 3]
```diff
@@ -1,11 +1,30 @@
...
+		height: 30px;
+		margin: var(--ee-margin-micro);
+		min-width: 30px;
+		padding: 4px;
+		width: 30px;
+	}
+
+	&__icon {
+		height: 16px;
+		margin: 0;
```
Comment (nicolad) [inline assets/src/application/ui/display/TimezoneTimeInfo/style.scss:18]: doesn't seem to have any effects, is this fixing smth?

File: assets/src/application/ui/display/TimezoneTimeInfo/Trigger.tsx [hunk 4]
```diff
@@ -8,19 +8,17 @@ interface TriggerProps {
...
-		/>
-	)
-);
+const Trigger = forwardRef<typeof IconButton, TriggerProps>(({ tooltip, ...props }, ref) => (
+	<IconButton
+		{...props}
+		borderless
+		color='white'
+		className='ee-timezone-info__button'
+		icon={() => <GlobalOutlined className='ee-timezone-info__icon' />}
```
Comment (nicolad) [inline assets/src/application/ui/display/TimezoneTimeInfo/Trigger.tsx]: I think it's better to lift these kind of styles at higher leve, had created `withSize` HOC, if any more props will be needed, can rename this differently.

File: assets/src/application/ui/display/withLabel/withLabel.tsx [hunk 5]
```diff
@@ -17,20 +17,21 @@ const withLabel = <P extends withLabelProps>(WrappedComponent: React.ComponentTy
 		...props
 	}) => {
 		const className = classNames(
-			labelClassName,
-			'ee-input-label__wrapper',
-			labelPosition && `ee-input-label__wrapper--${labelPosition}`
+			'ee-input__wrapper',
+			label && labelClassName,
```
[the remark this answers is not in the document]
Comment (nicolad) [reply assets/src/application/ui/display/withLabel/withLabel.tsx:21]: or I might miss smth here: `label && labelClassName` ?

File: assets/src/application/ui/forms/espressoForm/renderers/FieldRenderer.tsx (hunk 1 above)
Comment (nicolad) [reply assets/src/application/ui/forms/espressoForm/renderers/FieldRenderer.tsx:42]: it's ok, but maybe we can centralize this with a prop and apply needed styles at higher level,
the intention being a less fragile code

File: assets/src/application/ui/forms/espressoForm/styles.scss [hunk 6]
```diff
@@ -75,16 +75,17 @@
...
-			width: 50%;
-			display: inline-flex;
-			padding-right: var(--ee-margin-default);
-			margin: var(--ee-margin-default) 0;
-			label {
-				display: flex;
-				align-items: center;
+
+			&__help-info {
+				color: var(--ee-color-primary-low-contrast);
```
Comment (nicolad) [inline assets/src/application/ui/forms/espressoForm/styles.scss:87]: IMO, it's better to centralize color, to make it more consistent with other icons styles;
when doing previous refactor I noticed that is a bit hard to maintain these kinds of local patched styles

File: assets/src/infrastructure/ui/layout/menu/MenuToggle.tsx [hunk 7]
```diff
@@ -6,6 +6,7 @@ import { MenuToggleProps } from './types';
 const MenuToggle = React.forwardRef<typeof ChakraMenuButton, MenuToggleProps>(
 	({ children, variant = 'unstyled', ...props }, ref) => {
 		return (
+			// @ts-ignore
 			<ChakraMenuButton {...props} variant={variant} ref={ref}>
```
Comment (manzoorwanijk) [inline assets/src/infrastructure/ui/layout/menu/MenuToggle.tsx]: Better to add it only for that prop in a separate line
```suggestion
			<ChakraMenuButton
				{...props}
				// @ts-ignore
				variant={variant}
				ref={ref}
			>
```

Comment (nicolad) [reply assets/src/infrastructure/ui/layout/menu/MenuToggle.tsx]: good catch, updated :+1:

File: assets/src/infrastructure/ui/layout/menu/Menu.tsx (hunk 2 above)
Comment (nicolad) [reply assets/src/infrastructure/ui/layout/menu/Menu.tsx]: @manzoorwanijk I had reverted this

File: assets/src/infrastructure/ui/layout/menu/MenuToggle.tsx (hunk 7 above)
Comment (tn3rb) [reply assets/src/infrastructure/ui/layout/menu/MenuToggle.tsx]: thnx guys

File: assets/src/application/ui/forms/espressoForm/renderers/FieldRenderer.tsx (hunk 1 above)
Comment (tn3rb) [reply assets/src/application/ui/forms/espressoForm/renderers/FieldRenderer.tsx:42]: would be nice if our icons had a class applied to them by default, ex: `ee-icon--help-info`

Comment (nicolad) [reply assets/src/application/ui/forms/espressoForm/renderers/FieldRenderer.tsx:42]: yes, a generated one based on the component name would be smth interesting

File: assets/src/application/ui/display/TimezoneTimeInfo/style.scss (hunk 3 above)
Comment (tn3rb) [reply assets/src/application/ui/display/TimezoneTimeInfo/style.scss]: there _was_ a margin being applied to that icon that this was removing... ???

File: assets/src/application/ui/display/withLabel/withLabel.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (tn3rb) [reply assets/src/application/ui/display/withLabel/withLabel.tsx:21]: no you're right... i didn't realize that about that lib

File: assets/src/application/ui/forms/espressoForm/styles.scss (hunk 6 above)
Comment (tn3rb) [reply assets/src/application/ui/forms/espressoForm/styles.scss:87]: hmmm  ya this is case is different than the others since it is on the coloured background of the sidebar, but you're right that that could be encoded somehow maybe with an `inverse` prop or something

File: assets/src/application/ui/display/icons/svgs/withSize.tsx [hunk 8]
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+import classNames from 'classnames';
+import { withSizeProps, IconSize } from './types';
+import './style.scss';
+
+const withSize = <P extends withSizeProps>(WrappedComponent: React.ComponentType<P>) => {
+	type ISize = { size: IconSize };
+
+	const WithSize: React.FC<P & ISize> = ({ size, ...props }) => {
+		const className = classNames(props.className, size && `ee__icon--${size}`);
```
Comment (tn3rb) [inline assets/src/application/ui/display/icons/svgs/withSize.tsx:10]: can this not be baked right into our icon component?
or will that not work because those are dynamically generated?

I think of HOCs as enhancing a component with additional logic that doesn't belong in the component, whereas an icon's size seems like it should be a property of the icon itself, but there might be a technical reason why you have done this.

File: assets/src/infrastructure/ui/layout/menu/Menu.tsx (hunk 2 above)
Comment (tn3rb) [reply assets/src/infrastructure/ui/layout/menu/Menu.tsx]: I moved the tooltip out of the dropdown menu and then wrapped that component in the `witTooltip` HOC:

https://github.com/eventespresso/event-espresso-core/pull/2838/commits/c5e220e23329a6160d484282c88897089125ad0e#diff-1b4e6999f94ed2e1742f9f0aff62127eR8-R33

so that might have caused this change, but we should jump back to a commit before the above was merged to confirm

File: assets/src/application/ui/forms/espressoForm/renderers/FieldRenderer.tsx [hunk 9]
```diff
@@ -39,7 +39,7 @@ const FieldRenderer: React.FC<FieldRendererProps> = (props) => {
 				{label}
 				{info && (
 					<Tooltip placement='right' tooltip={info}>
-						<InfoCircleOutlined className='ee-form-item__help-info' />
+						<InfoCircleOutlined />
```
Comment (nicolad) [inline assets/src/application/ui/forms/espressoForm/renderers/FieldRenderer.tsx:42]: image

File: assets/src/application/ui/forms/espressoForm/styles.scss (hunk 6 above)
Comment (nicolad) [reply assets/src/application/ui/forms/espressoForm/styles.scss:87]: yes, `inverse` prop sounds like an interesting approach

File: assets/src/application/ui/display/icons/svgs/withEnhance.tsx [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+		.join('-');
+};
+
+const withEnhance = <P extends withSizeProps>(WrappedComponent: React.ComponentType<P>) => {
+	type ISize = { noMargin?: boolean; size: IconSize };
+
+	// @ts-ignore
+	const displayName = WrappedComponent?.render?.name;
+
+	const generatedIconClassName = displayName && 'ee-icon--' + camelToSnakeCase(displayName.replace('Svg', ''));
```
Comment (tn3rb) [inline assets/src/application/ui/display/icons/svgs/withEnhance.tsx:19]: cool 👍

File: assets/src/application/ui/display/icons/svgs/withEnhance.tsx [hunk 11]
```diff
@@ -0,0 +1,38 @@
+import React from 'react';
+import classNames from 'classnames';
+import { withSizeProps, IconSize } from './types';
+import './style.scss';
+
+const camelToSnakeCase = (str: string) => {
```
Comment (nicolad) [inline assets/src/application/ui/display/icons/svgs/withEnhance.tsx:6]: @tn3rb let me know if this is smth that can be reused, such that it can be moved into a cross-cutting folder

File: assets/src/application/ui/forms/espressoForm/renderers/FieldRenderer.tsx (hunk 1 above)
Comment (nicolad) [reply assets/src/application/ui/forms/espressoForm/renderers/FieldRenderer.tsx]: done

File: assets/src/application/ui/display/icons/svgs/withEnhance.tsx [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+	const WithSize: React.FC<P & ISize> = ({ noMargin, size, ...props }) => {
+		const className = classNames(
+			'ee-svg',
+			size && `ee__icon--${size}`,
+			noMargin && 'ee__icon--no-margin',
+			generatedIconClassName,
+			props.className
+		);
+
+		console.log('props.className', props.className);
```
Comment (tn3rb) [inline assets/src/application/ui/display/icons/svgs/withEnhance.tsx:30]: oops... need to remove this...

File: assets/src/application/ui/display/icons/svgs/withEnhance.tsx [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+	const generatedIconClassName = displayName && 'ee-icon--' + camelToSnakeCase(displayName.replace('Svg', ''));
+
+	const WithSize: React.FC<P & ISize> = ({ noMargin, size, ...props }) => {
+		const className = classNames(
+			'ee-svg',
+			size && `ee__icon--${size}`,
+			noMargin && 'ee__icon--no-margin',
+			generatedIconClassName,
+			props.className
+		);
```
Comment (tn3rb) [inline assets/src/application/ui/display/icons/svgs/withEnhance.tsx:28]: HOCception !!!
can this not just be combined with the logic inside `withEnhance` ???

File: assets/src/application/ui/display/withLabel/withLabel.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (nicolad) [reply assets/src/application/ui/display/withLabel/withLabel.tsx:21]: ok, will update that

**Comment by tn3rb** [comment]
> let me know if this is smth that can be reused

oh ya I think so... maybe extract to `/application/services/utilities/text/`

File: assets/src/application/ui/display/icons/svgs/withSize.tsx (hunk 8 above)
Comment (nicolad) [reply assets/src/application/ui/display/icons/svgs/withSize.tsx]: yes, we can pass directly height and width:
```
<GlobalOutlined heght="1rem" noMargin width="1rem"/>
```
but I thought that having an universal approach is easier to maintain, rather than explicit settings

File: assets/src/application/ui/display/icons/svgs/withEnhance.tsx (hunk 13 above)
Comment (nicolad) [reply assets/src/application/ui/display/icons/svgs/withEnhance.tsx:28]: splitted, thx

File: assets/src/application/ui/display/icons/svgs/withEnhance.tsx (hunk 12 above)
Comment (nicolad) [reply assets/src/application/ui/display/icons/svgs/withEnhance.tsx]: thx

File: assets/src/application/ui/display/TimezoneTimeInfo/style.scss (hunk 3 above)
Comment (nicolad) [reply assets/src/application/ui/display/TimezoneTimeInfo/style.scss]: you are right, that has been applied via `ee-svg`, my bad

File: assets/src/application/ui/display/icons/svgs/index.ts [hunk 14]
```diff
@@ -43,6 +47,6 @@ export { ReactComponent as WarningTriangle } from './warningTriangle.svg';
 
 // Enhanced icons
 
-export const InfoCircleOutlined = withEnhance(InfoCircleOutlinedSVG);
+export const InfoCircleOutlined = pipe(withGeneratedClassName, withSize)(InfoCircleOutlinedSVG);
 
-export const GlobalOutlined = withEnhance(GlobalOutlinedSVG);
+export const GlobalOutlined = pipe(withGeneratedClassName, withSize)(GlobalOutlinedSVG);
```
Comment (nicolad) [inline assets/src/application/ui/display/icons/svgs/index.ts]: the order of HOCs that are applied matters, tried the other way and it wasn't working as expected

File: assets/src/application/ui/display/icons/svgs/index.ts [hunk 15]
```diff
@@ -1,4 +1,8 @@
-import withEnhance fro./withEnhancenhance';
+import { pipe } from 'ramda';
```
Comment (nicolad) [inline assets/src/application/ui/display/icons/svgs/index.ts:1]: I know only this function from `ramda`

File: assets/src/application/ui/display/icons/svgs/index.ts [hunk 16]
```diff
@@ -39,3 +44,9 @@ export { ReactComponent as UpCircleFilled } from './upCircleFilled.svg';
 export { ReactComponent as Warning } from './warning.svg';
 
 export { ReactComponent as WarningTriangle } from './warningTriangle.svg';
+
+// Enhanced icons
+
+export const InfoCircleOutlined = pipe(withGeneratedClassName, withSize)(InfoCircleOutlinedSVG);
+
+export const GlobalOutlined = pipe(withGeneratedClassName, withSize)(GlobalOutlinedSVG);
```
Comment (tn3rb) [inline assets/src/application/ui/display/icons/svgs/index.ts:52]: sry I don't think I was very clear in my comments before...

when I said: 

> can this not just be combined with the logic inside withEnhance ???

what I meant was that having a separate HOC **just** for applying the icon name as a css class to an icon...
as well as **another** separate HOC for applying size and margin related classes to an icon...
seems kind of redundant...

They are both just applying classes to an icon, therefore it seems more logical to me to combine these together into one HOC that applies icon classes.
To go further I **actually** think that typical modifier classes for a component should be part of the component itself and not an HOC which imho should be reserved for applying unrelated and/or atypical changes to a component.

In both of these cases however I think that the enhancements should apply to ALL of our icons:

- adding the icon name as a css class should be applied to ALL of our icons, because that will undoubtedly be handy and allow any icon to be targeted via css. This is similar to how all of the WP dashicons have their icon name as part of their css classes like `dashicons dashicons-calendar`.

- being able to change the size of an icon also seems like something that all icons should have the capacity to do. AFAIK this can be achieved indirectly by modifying the font size for the container, but imho it owul dbe better to have some standardized sizes that could be controlled more centrally

- removing margins also seems pretty generic and could likely be a nice to have for any icon

So bottom line, let's just have a single HOC for adding css modifiers
(and I'm assuming that the HOC is used because the icon components are dynamically generated and there isn't a single `Icon` component where that logic can be centralized?

File: assets/src/application/ui/forms/espressoForm/renderers/FieldRenderer.tsx (hunk 9 above)
Comment (tn3rb) [reply assets/src/application/ui/forms/espressoForm/renderers/FieldRenderer.tsx:42]: love it 👍 great job

File: assets/src/application/ui/display/icons/svgs/index.ts (hunk 16 above)
Comment (nicolad) [reply assets/src/application/ui/display/icons/svgs/index.ts:52]: there are custom templates from `svgr` which would apply that HOC to each of the SVG component: https://react-svgr.com/docs/custom-templates/
***
need to think of an easier way to achieve this

Comment (nicolad) [reply assets/src/application/ui/display/icons/svgs/index.ts:52]: tried dynamically to apply those HOC, but got no luck:
```
import { pipe } from 'ramda';

import withGeneratedClassName from './svgs/withGeneratedClassName';
import withSize from './svgs/withEnhance';

import * as icons from './svgs';

const ehancedIcons = Object.entries(icons).reduce((acc, [name, component]) => {
	return (acc[name] = pipe(withGeneratedClassName, withSize)(component));
}, {});

export default ehancedIcons;
```

File: assets/src/application/ui/display/icons/svgs/withEnhance.tsx [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+
+		return <WrappedComponent {...(props as P)} className={className} />;
+	};
+
+	return WithEnhance;
+};
```
Comment (tn3rb) [inline assets/src/application/ui/display/icons/svgs/withEnhance.tsx:30]: ya that's so much better imho
